### ISSUE TITLE
job summary subsection: per-module i/o characteristics

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#465]
Once we have a functional template we can build on, we should add a subsection related to I/O characteristics for different modules contained in the log. There are a number of characteristics that we have traditionally captured for a few Darshan modules (specifically, POSIX, MPI-IO, and STDIO), but it would likely be straightforward to extend the idea to other modules that aren't currently covered (HDF5 being an obvious one). 

I think as a starting point, we could try to reproduce the following types of graphs (note numbers at top) from our current job summary tool, probably starting with just POSIX and MPI-IO for now:
![image](https://user-images.githubusercontent.com/24571836/131564549-b49ff6eb-2e1a-418f-85f5-e9e605d5b5ab.png)
 
1. I/O cost graphs: Provide estimates on per-process I/O time spent doing read/write/metadata operations for the corresponding module API (POSIX, MPI-IO, etc.). This requires us re-implementing some performance estimate logic that is currently done in our darshan-parser C utility. Display using stacked bar graph. Important to display cost both in terms of percentage runtime and elapsed runtime in seconds.
2. I/O operation count graphs: Provide total number of different I/O operations executed using the module API. Things like opens, reads, writes, stats, seeks, etc. To generalize this plot, likely need an API here to allow modules to specify what operations they support and what counters are needed to access them (e.g., mapping "opens" to `POSIX_OPENS` counter for the POSIX module). Display using a bar graph, with each operation on the x-axis. Likely need log-scale support on y-axis for large ranges.
3. Access size histograms: Provide distributions of all read and write operation access sizes using histograms. A few modules capture this information in the same way (e.g., `POSIX_SIZE_WRITE_0_100`-`POSIX_SIZE_WRITE_1G_PLUS` counters for the POSIX module) so should be generalizable. 
4. Common access sizes: Darshan does capture information on the 4 most common access sizes issued to each file, so we could aggregate that information and summarize the most common ones for useful details that are too fine-grained to see in histogram described for plot 3.

The template engine would need to detect the presence of these modules, then generate the figures described above along with corresponding HTML code in different module sub-sections of the report as shown.

After completing this, we could experiment with extending these sorts of summaries to STDIO and HDF5 modules, and could also look at adding module-specific summaries in addition to the "common" ones described here. An example module-specific graph would be the figure in our current job summary tool that displays percentage of POSIX operations that are sequential or consecutive in files.

**Comment by nawtrey** [comment]
I'm gonna take a stab at the the I/O cost graph. After experimenting a bit it looks like I can manipulate the `fcounter` dataframe and generate the stacked bar graph with `pandas.dataframe.plot.bar()`, so there shouldn't be much custom calculation or plotting code (fingers crossed).

**Comment by carns** [comment]
Just need to completed item number 2 for the milestone (op counts graph)

**Comment by nawtrey** [comment]
Once #654 is merged, I think the only remaining tasks not completed (or covered in other issues) are:
1. log scaling for the "I/O Operation Counts" figure
2. general `STDIO` support

I think I'll open separate issues for these, and if we want to include them in a milestone we can.

**Comment by tylerjereddy** [comment]
Since the old perl reports are consistent with your observation related to point `2` and a separate issue is open for point `1`, I'll close this with gh-654 merged, and if Phil or Shane disagree they can reopen.
